src/anyio/_core/_fileio.py

```diff
@@ -114,3 +125,346 @@ async def open_file(file: Union[str, PathLike, int], mode: str = 'r', buffering:
     fp = await to_thread.run_sync(open, file, mode, buffering, encoding, errors, newline,
                                   closefd, opener)
     return AsyncFile(fp)
+
+
+@dataclass(eq=False)
+class _PathIterator(AsyncIterator['Path']):
+    iterator: Iterator[PathLike]
+
+    async def __anext__(self) -> 'Path':
+        nextval = await to_thread.run_sync(next, self.iterator, None, cancellable=True)
+        if nextval is None:
+            raise StopAsyncIteration from None
+
+        return Path(cast(PathLike, nextval))
+
+
+class Path:
+    """
+    An asynchronous version of :class:`pathlib.Path`.
+
+    This class cannot be substituted for :class:`pathlib.Path` or :class:`pathlib.PurePath`, but
+    it is compatible with the :class:`os.PathLike` interface.
+
+    It implements the Python 3.10 version of :class:`pathlib.Path` interface, except for the
+    deprecated :meth:`~pathlib.Path.link_to` method.
+
+    Any methods that do disk I/O need to be awaited on. These methods are:
+
+    * :meth:`~pathlib.Path.absolute`
+    * :meth:`~pathlib.Path.chmod`
+    * :meth:`~pathlib.Path.cwd`
+    * :meth:`~pathlib.Path.exists`
+    * :meth:`~pathlib.Path.expanduser`
+    * :meth:`~pathlib.Path.group`
+    * :meth:`~pathlib.Path.hardlink_to`
+    * :meth:`~pathlib.Path.home`
+    * :meth:`~pathlib.Path.is_block_device`
+    * :meth:`~pathlib.Path.is_char_device`
+    * :meth:`~pathlib.Path.is_dir`
+    * :meth:`~pathlib.Path.is_fifo`
+    * :meth:`~pathlib.Path.is_file`
+    * :meth:`~pathlib.Path.is_mount`
+    * :meth:`~pathlib.Path.lchmod`
+    * :meth:`~pathlib.Path.lstat`
+    * :meth:`~pathlib.Path.mkdir`
+    * :meth:`~pathlib.Path.open`
+    * :meth:`~pathlib.Path.owner`
+    * :meth:`~pathlib.Path.read_bytes`
+    * :meth:`~pathlib.Path.read_text`
+    * :meth:`~pathlib.Path.readlink`
+    * :meth:`~pathlib.Path.rename`
+    * :meth:`~pathlib.Path.replace`
+    * :meth:`~pathlib.Path.rmdir`
+    * :meth:`~pathlib.Path.samefile`
+    * :meth:`~pathlib.Path.stat`
+    * :meth:`~pathlib.Path.touch`
+    * :meth:`~pathlib.Path.unlink`
+    * :meth:`~pathlib.Path.write_bytes`
+    * :meth:`~pathlib.Path.write_text`
+
+    Additionally, the following methods return an async iterator yielding :class:`~.Path` objects:
+
+    * :meth:`~pathlib.Path.glob`
+    * :meth:`~pathlib.Path.iterdir`
+    * :meth:`~pathlib.Path.rglob`
+    """
+
+    __slots__ = '_path'
+
+    def __init__(self, *args: Union[str, PathLike]) -> None:
+        self._path: Final[pathlib.Path] = pathlib.Path(*args)
+
+    def __fspath__(self) -> str:
+        return self._path.__fspath__()
+
+    def __str__(self) -> str:
+        return self._path.__str__()
```

Comment:
These dunders need to be called via the class with `type(v).__str__(v)` or just `str(v)`

Reply:
I don't understand this comment. `str(x)` or `x.__str__()` or `type(x).__str__(x)` are all equivalent.

Reply:
@smurfix they're different:

```python
class Foo:
    def __init__(self) -> None:
        self.__str__ = lambda: "x.__str__()"

    def __str__(self) -> str:
        return "type(x).__str__(x)"

print(f"{str(Foo())=}")  # str(Foo())='type(x).__str__(x)'
print(f"{Foo().__str__()=}")   # Foo().__str__()='x.__str__()'
```

Reply:
Yes I know that `str()` takes a shortcut and doesn't do a lookup on the instance. But which real-world class assigns to `self.__str__`? `pathlib.Path`certainly doesn't.
```
>>> import pathlib as p
>>> x=p.Path("a/b/c")
>>> x
PosixPath('a/b/c')
>>> str(x)
'a/b/c'
>>> x.__str__()
'a/b/c'
>>> 
>>> type(x).__str__(x)
'a/b/c'
>>>
```

Reply:
I'm also not understanding why this needs to be changed. Is there a potential problem with this way of doing things?

Reply:
Ah I forgot pathlib.Path was frozen. It's usually a problem but not in this specific case